### ISSUE TITLE
Possible multi-partitioning happening when partitions(1) for read queries

### ISSUE THREAD

**Issue by moxious** [neo4j-contrib/neo4j-spark-connector#337]
Reported by a customer but also replicated by me.

```
# Create a GDS in-memory graph for authors.

cypher = """
  CALL gds.graph.create.cypher(
    'authors', 
    'MATCH (a:Author) RETURN id(a) as id',
    'MATCH (a1:Author)-[r1:AUTHOR]->(p:Paper)<-[r2:AUTHOR]-(a2:Author)
     WHERE 
       id(a1) < id(a2) AND 
       r1.index <= 4 AND 
       r2.index <=4 
     RETURN id(a1) as source, id(a2) as target, count(p) as weight
    ',
    {}
  ) YIELD graphName RETURN graphName
"""

spark.read.format("org.neo4j.spark.DataSource") \
  .format("org.neo4j.spark.DataSource") \
  .option("url", dbutils.widgets.get("url")) \
  .option("authentication.basic.username", dbutils.widgets.get("user")) \
  .option("authentication.basic.password", dbutils.widgets.get("password")) \
  .option("query", cypher) \
  .option("partitions", "1") \
  .load().show()
```

The error is this: 

```
org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 30.0 failed 4 times, most recent failure: Lost task 0.3 in stage 30.0 (TID 166) (10.57.233.165 executor 0): org.neo4j.driver.exceptions.ClientException: Failed to invoke procedure `gds.graph.create.cypher`: Caused by: java.lang.IllegalArgumentException: A graph with name &#39;authors&#39; already exists.
```

Now, when you try to create duplicate graphs in GDS this error is predictable.   Note the use of `.options("partitions", "1")`.

We need to double check on this.  The error itself is OK and doesn't cause harm because the query does create the GDS virtual graph.  But this error suggests that something is being executed in parallel when we tell the connector to use a single partition.

Tested using Databricks runtime 8.2 which includes Spark 3.1.1 -- it's possible this may be related?

**Comment by moxious** [comment]
@utnaf found this reference indicating that single partition should be single task and shouldn't be multi-threaded.  We know of no reason why the underlying driver would multi-thread either (although each driver instance does maintain a connection pool) https://stackoverflow.com/questions/47782099/what-is-the-relationship-between-tasks-and-partitions

**Comment by utnaf** [comment]
At the current status this behaviour is "honest". Meaning that, for read operations, we first execute the query to extract the schema of the DataFrame, then we execute the query to get the data, and here happen the second execution.

At the moment there is no workaround. The solution is to implement the user specified schema support for our DataSource (available for both Spark 2.4 and Spark 3). This will allow the user to specify the schema, letting us skip the schema generation that causes the additional query execution.

Example of the `.schema()` method

```scala
val df = ss.read.format(classOf[DataSource].getName)
      .schema(StructType(Seq(StructField("name", DataTypes.StringType))))
      .option("url", SparkConnectorScalaSuiteIT.server.getBoltUrl)
      .option("labels", "Person")
      .load()
```

**Comment by moxious** [comment]
Notes for later:

- Spark is tightly schema bound so we have to determine shape of resulting schema
- If, in order to do this in Neo4j, we have to execute the query this is going to make running GDS queries via the connector not the best approach as some of them will always be extremely expensive (for example running a big graph compute)
- Partially because of this I'm spiking [gds-python](https://github.com/moxious/gds-python) in the background as an experiment for pure python GDS in Spark notebooks
